src/main/java/org/folio/circulation/services/LostItemFeeChargingService.java

```diff
@@ -54,35 +71,108 @@ public LostItemFeeChargingService(Clients clients) {
     this.locationRepository = LocationRepository.using(clients);
     this.eventPublisher = new EventPublisher(clients.pubSubPublishingService());
     this.feeFineScheduledNoticeService = FeeFineScheduledNoticeService.using(clients);
+    this.refundService = new LostItemFeeRefundService(clients);
+    this.accountRepository = new AccountRepository(clients);
+    this.feeFineService = new FeeFineService(clients);
+    this.userRepository = new UserRepository(clients);
   }
 
   public CompletableFuture<Result<Loan>> chargeLostItemFees(
     Loan loan, DeclareItemLostRequest request, String staffUserId) {
+    this.userId = staffUserId;
+    this.servicePointId = request.getServicePointId();
 
     final ReferenceDataContext referenceDataContext = new ReferenceDataContext(
       loan, request, staffUserId);
-
     return lostItemPolicyRepository.getLostItemPolicyById(loan.getLostItemPolicyId())
       .thenApply(result -> result.map(referenceDataContext::withLostItemPolicy))
       .thenCompose(refDataResult -> refDataResult.after(referenceData -> {
-        if (shouldCloseLoan(referenceData.lostItemPolicy)) {
-          log.debug("Loan [{}] can be closed because no fee will be charged", loan.getId());
-          return closeLoanAndUpdateInStorage(loan)
-            .thenCompose(r -> r.after(eventPublisher::publishClosedLoanEvent))
-            .thenApply(r -> r.map(v -> loan));
+        log.info("Checking for existing lost item fees for loan [{}]", loan.getId());
+        return accountRepository.findAccountsForLoan(loan)
+          .thenCompose(result -> {
+            this.loanWithAccountData = result.value();
+
+            if (!hasLostItemFees(loanWithAccountData) && !shouldCloseLoan(referenceData.lostItemPolicy)) {
+              log.info("No existing lost item fees found, applying lost item fees to loan [{}]", loan.getId());
+              return applyFees(referenceData, loanWithAccountData);
+            }
+
+            if (!hasLostItemFees(loanWithAccountData) && shouldCloseLoan(referenceData.lostItemPolicy)) {
+              log.info("No existing lost item fees found, losing loan [{}] as lost and paid.", loan.getId());
+              return CloseLoanAsLostandPaidAndPublishEvent(loan);
+            }
+            log.info("Existing lost item fees found for loan [{}], trying to clear", loan.getId());
+            return removeAndRefundFees(userId, servicePointId, loanWithAccountData)
+              .thenCompose(refundResult -> {
+                if (refundResult.failed()) {
+                  log.error("Unable to clear lost item fees for loan [{}], aborting charge fees.", loan.getId());
+                  return CompletableFuture.completedFuture(failed(refundResult.cause()));
+                }
```

Comment:
That's part of it, but more because if there is a problem, I want the code to exit here in case of failure and not go on to mark the loan as lost and paid or apply fees-in this situation, the loan should just be marked as declared lost and any remaining fees should be left.  Is there a better way to do that?  I'm still very new to promises, so it's entirely possible there's something I'm missing here.

Reply:
> That's part of it, but more because if there is a problem, I want the code to exit here in case of failure and not go on to mark the loan as lost and paid or apply fees-in this situation, the loan should just be marked as declared lost and any remaining fees should be left. 

Thank you for explaining the intent

> Is there a better way to do that? I'm still very new to promises, so it's entirely possible there's something I'm missing here.

The confusion is likely at least in part because the `result` is not part of Java's future model. It is something I built to deal with the complexity of managing error flow when composing many asynchronous operations. 

I wrote it at a time when callbacks were popular in FOLIO and complex flows were in callback hell. I didn't understand `CompletableFuture` very well at the time. Vert.x has moved on since then as well. Part of this is that I detest flow of control using exceptions, which may have twisted this design somewhat.

The methods in the `Result` class will typically forward an existing failure and skip whatever processing is has been asked to do with a success. If this is the goal, I think the code can be simplified. If I get chance, I'll raise a follow up pull request that demonstrates this. Please don't wait for this to merge this code.

Reply:
Merging now.